rpcs3/Emu/Cell/Modules/cellAtracXdec.h

```diff
@@ -0,0 +1,300 @@
+
+	be_t<u32> run_thread_mutex; // sys_mutex_t
+	be_t<u32> run_thread_cond;  // sys_cond_t
+	be_t<u32> run_thread = true;
+
+	const AdecCb<AdecNotifyAuDone> notify_au_done;
+	const AdecCb<AdecNotifyPcmOut> notify_pcm_out;
+	const AdecCb<AdecNotifyError> notify_error;
+	const AdecCb<AdecNotifySeqDone> notify_seq_done;
+
+	const vm::bptr<u8> work_mem;
+
+	// HLE exclusive
+	u64 cmd_counter = 0;             // For debugging
+	AtracXdecCmd cmd;                // For savestates; if savestate was created while processing a decode command, we need to save the current command
+	b8 skip_getting_command = false; // For savestates; skips getting a new command from the queue
+	b8 skip_next_frame;              // Needed to emulate behavior of LLE SPU program, it doesn't output the first frame after a sequence reset or error
+
+	u8 spurs_stuff[58]; // 120 bytes on LLE, pointers to CellSpurs, CellSpursTaskset, etc.
+
+	CellAdecParamAtracX atracx_param;
+
+	u8 reserved;
+	b8 first_decode;
+
+	AtracXdecDecoder decoder;
+
+	AtracXdecContext(vm::ptr<AdecNotifyAuDone> notifyAuDone, vm::ptr<void> notifyAuDoneArg, vm::ptr<AdecNotifyPcmOut> notifyPcmOut, vm::ptr<void> notifyPcmOutArg,
+		vm::ptr<AdecNotifyError> notifyError, vm::ptr<void> notifyErrorArg, vm::ptr<AdecNotifySeqDone> notifySeqDone, vm::ptr<void> notifySeqDoneArg, vm::bptr<u8> work_mem)
+		: notify_au_done{ notifyAuDone, notifyAuDoneArg }
+		, notify_pcm_out{ notifyPcmOut, notifyPcmOutArg }
+		, notify_error{ notifyError, notifyErrorArg }
+		, notify_seq_done{ notifySeqDone, notifySeqDoneArg }
+		, work_mem(work_mem)
+		, decoder(work_mem)
+	{
+	}
+
+	void exec(ppu_thread& ppu);
+
+	template <AtracXdecCmdType type>
+	error_code send_command(ppu_thread& ppu, auto&&... args);
+};
+
+static_assert(std::is_standard_layout_v<AtracXdecContext>);
+CHECK_SIZE_ALIGN(AtracXdecContext, 0x268, 8);
+
+constexpr u32 ATXDEC_SPURS_STRUCTS_SIZE = 0x1cf00; // CellSpurs, CellSpursTaskset, context, etc.
+constexpr u16 ATXDEC_SAMPLES_PER_FRAME = 0x800;
+constexpr u16 ATXDEC_MAX_FRAME_LENGTH = 0x2000;
+constexpr std::array<u8, 8> ATXDEC_NCH_BLOCKS_MAP = { 0, 1, 1, 2, 3, 4, 5, 5 };
+
+// Expected output channel order
+//  - for 1 to 7 channels: Front Left, Center, Front Right, Rear Left, Rear Right, Rear Center, LFE
+//  - for 8 channels:      Front Left, Front Right, Center, LFE, Rear Left, Rear Right, Side Left, Side Right
+// FFmpeg output
+//  - ver <= 5.1.2: Front Left, Front Right, Center, Rear Left, Rear Right, Rear Center, Side Left, Side Right, LFE
+//  - ver >= 5.1.3: Front Left, Front Right, Center, LFE, Rear Left, Rear Right, Rear Center, Side Left, Side Right
+constexpr u8 ATXDEC_AVCODEC_CH_MAP[7][8] =
+{
+	{ 0 },
+	{ 0, 1 },
+	{ 0, 2, 1 },
+	{ 0, 2, 1, 3 },
+#if LIBAVCODEC_VERSION_INT < AV_VERSION_INT(59, 51, 101)
+	{ 0, 2, 1, 3, 4, 5 },
+	{ 0, 2, 1, 3, 4, 5, 6 },
+	{ 0, 1, 2, 4, 5, 6, 7, 3 }
+#else
+	{ 0, 2, 1, 4, 5, 3 },
+	{ 0, 2, 1, 4, 5, 6, 3 },
+	{ 0, 1, 2, 3, 4, 5, 6, 7 }
+#endif
+};
```

Comment:
It's not the enum, it's the index of each channel in FFmpeg's output.

The version check is there because the channel order FFmpeg outputs changes in this commit:
https://github.com/FFmpeg/FFmpeg/commit/3819719099df601c470e961b9d49b9100c65641b
The check makes sure the output of cellAtracXdec is still correct even when we update FFmpeg or someone uses their own local library.